File: src/Rcpp/ThetaLog/ThetaLog.R
setwd('src/Rcpp/ThetaLog')
library(TMB)
compile("ThetaLog.cpp", flags= "-w")
dyn.load(dynlib("ThetaLog"))

## Read data
Y <- scan("ThetaLog.dat", skip=3, quiet=TRUE)
data <- list(Y=Y)

## Parameter initial guess
parameters <- list(
  X = data$Y*0,
  logr0 = 0,
  logtheta = 0,
  logK = 6,
  logQ = 0,
  logR = 0
)

## Fit model
obj <- MakeADFun(data, parameters, random="X", DLL="ThetaLog")
newtonOption(obj, smartsearch=FALSE)
obj$fn()
obj$gr()
system.time(opt <- nlminb(obj$par, obj$fn, obj$gr))
rep <- sdreport(obj)
rep

File: src/Rcpp/ThetaLog/ThetaLog.cpp
// #include <TMB.hpp>
#include "model.hpp"


template<class Type>
Type objective_function<Type>::operator() ()
{
  ThetaLog<Type>* inst = ThetaLog<Type>::get_instance();
  
  /* Data section */
  DATA_VECTOR(Y);
  /* Parameter section */
  PARAMETER_VECTOR(X);
  PARAMETER(logr0);
  PARAMETER(logtheta);
  PARAMETER(logK);
  PARAMETER(logQ);
  PARAMETER(logR);
  
  inst->Y = Y;
  inst->X = X;
  inst->logr0 =logr0;
  inst->logtheta = logtheta;
  inst->logK = logK;
  inst->logQ = logQ;
  inst->logR = logR;
  
  
  return inst->evaluate();
}

